doc/developer/design/20230721_persist_fast_path.md

```diff
@@ -0,0 +1,83 @@
+
+- In `environmentd`, generalize the existing fast-path planning logic. As of today, fast-path plans were only generated when the queried collection has at least one index. We’ll extend this logic to also generate these plans for unindexed collections that are backed by a Persist shard.
+- When `clusterd` receives a peek against a Persist-backed collection — as opposed to an arrangement, which is always the case today — it will create a new `PersistPeek` operation that streams the data from a Persist shard. While it’s possible to efficiently and incrementally stream consolidated data from a Persist shard, no such API exists today; we’ll need to add one.
+
+## Detailed description
+
+### Planning
+
+Today, `environmentd` will generate a fast-path plan when a dataflow can be represented as: a `Get` from an indexed collection, a map-filter-project stage, and some “finishing” logic including order-by/limit. We’ll change this to also allow the `Get` to get from a non-indexed collection backed by a Persist shard.
+
+If we want to ensure that the cost of the query is proportional to the result set size and not the size of the collection, we'll need to introduce some additional restrictions:
+- No filters. Index-based fast-path queries are often used for single-key lookups, like `select * from indexed_collection where id = 'XXX'`. Persist cannot currently support efficient queries with this shape.
+- No `ORDER BY`. This would require us to load and sort the full collection, then apply the limit.
+
+We may be able to relax these requirements in the future. (For example, if we order the Persist data by the data's "logical" ordering and not its serialized representation, we could support primary-key lookups more efficiently.)
+
+Today, every select query generates a peek request into an arrangement. (Even non-fast-path queries work by creating a new dataflow that writes out its data into an arrangement, then peeking into that arrangement.) This design introduces a second type of peek: one that peeks directly into the backing Persist shard. Concrentely, this means that `environmentd` will send `ComputeCommand::Peek` commands that reference a persisted collection’s `Id`, not the id of an index or arrangement.
+
+### Peeking
+
+When `clusterd` receives a peek command, it stores and tracks a `PendingPeek` struct with the arrangement and request metadata in the compute state, polls until that arrangement has data for the requested timestamp, then read that data synchronously back out of the arrangement before returning a `ComputeResponse::Peek` with the results.
+
+Since Persist’s API is asynchronous, one straightforward approach would be to fork off a new task for every incoming Peek, and hold on to the task handle. We can periodically poll the task for completion, and cancel it if the peek itself is cancelled by the user.
```

Comment:
I am worried about our ability to debug these tasks in production. For dataflows we have plenty of ways to observe their execution through the introspection sources. For these peek tasks we would have to add new mechanisms, and maintain/improve them separately. I think we don't want to release this feature without observability, considering that it seems like one that can easily introduce performance issues (`SELECT * FROM megashard`).

Perhaps this would be less concerning if we decided to limit persist peeks to very small result sets, e.g., demand that they include a `LIMIT` below 10. (But at that point we might as well do the reads from `environmentd` directly.)

Reply:
> I think we don't want to release this feature without observability [...]

Extremely agreed!

> For dataflows we have plenty of ways to observe their execution through the introspection sources.

True! Though I feel like the most apt comparison is with the existing `PendingPeek` tasks, which have similar constraints to the proposed Persist-based ones, and `SELECT * FROM megashard` will similarly kick off a lot of work today if `megashard` is indexed. My hope was that by sharing some abstractions with the existing fast-path peek path we'd be able to share some of the existing observability for existing fast-path peeks.

Reply:
Peeks today are just scans/lookups of in-memory arrangements, so I'd think their potential for things going wrong should be much smaller than reading from persist over the network. But also I think we don't have much observability into how peek processing goes, aside from how long it takes and what the result is. That's perhaps something we should look at independently!

Reply:
We addressed this by updating the existing Peek observability to handle Persist peeks, plus also some Persist-side metrics to track some finer grained detail.